spd/models/component_model.py

```diff
@@ -510,6 +522,50 @@ def cache_hook(_: nn.Module, input: tuple[Tensor, "..."], param_name: str) -> No
             for module in self.components_or_modules.values():
                 module.make_pristine()
 
+    def _forward_with_pre_forward_cache_components_hooks(
+        self,
+        *args: Any,
+        mask_infos: dict[str, ComponentsMaskInfo],
+        module_names: list[str],
+        **kwargs: Any,
+    ) -> tuple[Any, dict[str, Tensor]]:
+        """Forward pass with component replacements and caching at the input to the modules given by `module_names`.
+
+        Args:
+            mask_infos: Dictionary mapping module names to ComponentMaskInfo
+            module_names: List of module names to cache the inputs to.
+
+        Returns:
+            Tuple of (model output, cache dictionary)
+        """
+        cache = {}
+        handles: list[RemovableHandle] = []
+
+        def cache_hook(_: nn.Module, input: tuple[Tensor, "..."], param_name: str) -> None:
+            cache[param_name] = input[0]
+
+        # Register hooks
+        for raw_module_name in module_names:
+            is_identity = raw_module_name.startswith("identity_")
+            module_name = (
+                raw_module_name.removeprefix("identity_") if is_identity else raw_module_name
+            )
+            module = self.patched_model.get_submodule(module_name)
+            assert module is not None, f"Module {module_name} not found"
+            handles.append(
+                module.register_forward_pre_hook(partial(cache_hook, param_name=raw_module_name))
+            )
+
+        # Set up component replacements
+        with self._replaced_modules(mask_infos=mask_infos):
+            try:
+                raw_out = self.patched_model(*args, **kwargs)
+                out = self._extract_output(raw_out)
+                return out, cache
+            finally:
+                for handle in handles:
+                    handle.remove()
+
```

Comment:
I think you can just directly use _forward_with_pre_forward_cache_hooks with minor modifications instead of creating this new method? Modifications:
- You can add a "mask_infos" argument to it, which can be None.
- Add the `with self._replaced_modules` context manager to the method. Pass in an empty dictionary when mask_infos is None. It should then be a no-op method (apart from calling make_pristine(), which is fine).

Everything else should be the same unless I'm missing something.